File: app/src/test/java/io/intrepid/contest/screens/sendinvitations/selectcontacts/SelectContactsPresenterTest.java
package io.intrepid.contest.screens.sendinvitations.selectcontacts;

import org.junit.Test;
import org.mockito.Mock;

import java.util.ArrayList;
import java.util.List;

import io.intrepid.contest.models.Contact;
import io.intrepid.contest.testutils.BasePresenterTest;

import static junit.framework.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.verify;

public class SelectContactsPresenterTest extends BasePresenterTest<SelectContactsPresenter> {
    @Mock
    SelectContactsContract.View mockView;

    private void setupContactSelectionScreen() {
        setup(true, new ArrayList<>());
    }

    private void setupPreviewContactsScreen() {
        setup(false, getMockContactList(true));
    }

    private void setup(boolean contactSelectionEnabled, List<Contact> contactList) {
        presenter = new SelectContactsPresenter(mockView,
                                                testConfiguration,
                                                contactSelectionEnabled,
                                                contactList);
    }

    @Test
    public void onViewCreatedShouldSetupAdapterWithContactSelectionWhenShowingContactSelectionScreen() {
        setupContactSelectionScreen();
        presenter.onViewCreated();
        verify(mockView).setupAdapter(true);
    }

    @Test
    public void onViewCreatedShouldSetupAdapterWithoutContactSelectionWhenShowingPreviewContactsScreen() {
        setupPreviewContactsScreen();
        presenter.onViewCreated();
        verify(mockView).setupAdapter(false);
    }

    @Test
    public void onViewCreatedShouldDisplayPhoneContactListWhenShowingContactSelectionScreen() {
        setupContactSelectionScreen();
        presenter.onViewCreated();
        verify(mockView).displayPhoneContactList();
    }

    @Test
    public void onViewCreatedShouldShowProgressBarWhenShowingContactSelectionScreen() {
        setupContactSelectionScreen();
        presenter.onViewCreated();
        verify(mockView).showProgressBar(true);
    }

    @Test
    public void onViewCreatedShouldUpdateAdapterContactListWhenShowingPreviewContactsScreen() {
        setupPreviewContactsScreen();
        presenter.onViewCreated();
        verify(mockView).updateAdapterContactList(any());
    }

    @Test
    public void onViewCreatedShouldHideProgressBarWhenShowingPreviewContactsScreen() {
        setupPreviewContactsScreen();
        presenter.onViewCreated();
        verify(mockView).showProgressBar(false);
    }

    @Test
    public void onCreateOptionsMenuShouldCreateMenuSearchItemWhenShowingContactSelectionScreen() {
        setupContactSelectionScreen();
        presenter.onCreateOptionsMenu();
        verify(mockView).createMenuSearchItem();
    }

    @Test
    public void onCreateOptionsMenuShouldNotCreateMenuSearchItemWhenShowingPreviewContactsScreen() {
        setupPreviewContactsScreen();
        presenter.onCreateOptionsMenu();
        verify(mockView, never()).createMenuSearchItem();
    }

    @Test
    public void onContactListUpdatedShouldHideProgressBar() {
        setupContactSelectionScreen();
        presenter.onContactListUpdated(getMockContactList(false));
        verify(mockView).showProgressBar(false);
    }

    @Test
    public void onContactListUpdatedShouldUpdateContactListFilteringOutContactsWithoutPhoneOrEmail() {
        setupContactSelectionScreen();
        List<Contact> initialContactList = getMockContactList(false);

        presenter.onContactListUpdated(initialContactList);

        verify(mockView).updateAdapterContactList(argThat(argument -> argument.size() ==
                initialContactList.size() - 1));
    }

    @Test
    public void onContactListUpdatedShouldUpdateAdapterContactListKeepingDataFromContactWhenInExistedInPreviousList() {
        // Initial contact list
        setupContactSelectionScreen();
        List<Contact> initialList = getMockContactList(false);
        initialList.get(0).setSelected(true);
        presenter.onContactListUpdated(initialList);
        reset(mockView);
        // Update list
        List<Contact> listAfterSearch = getMockContactList(false).subList(0, 1);
        listAfterSearch.get(0).setSelected(false);

        presenter.onContactListUpdated(listAfterSearch);

        verify(mockView).updateAdapterContactList(
                argThat(adapterList -> (adapterList.get(0) == initialList.get(0))));
    }

    @Test
    public void onQueryTextSubmitShouldAlwaysReturnTrue() {
        setupContactSelectionScreen();
        assertTrue(presenter.onQueryTextSubmit(""));
    }

    @Test
    public void onQueryTextChangeShouldUpdateContactSearchFilterWhenTextIsEmpty() {
        setupContactSelectionScreen();
        presenter.onQueryTextChange("");
        verify(mockView).updateContactSearchFilter(anyString());
    }

    @Test
    public void onQueryTextChangeShouldNotUpdateContactSearchFilterWhenTextHasOneCharacter() {
        setupContactSelectionScreen();
        presenter.onQueryTextChange("1");
        verify(mockView, never()).updateContactSearchFilter(anyString());
    }

    @Test
    public void onQueryTextChangeShouldUpdateContactSearchFilterWhenTextHasTwoCharactersOrMore() {
        setupContactSelectionScreen();
        presenter.onQueryTextChange("12");
        verify(mockView).updateContactSearchFilter(anyString());
    }

    @Test
    public void onQueryTextChangeShouldHideProgressBarWhenTextIsEmpty() {
        setupContactSelectionScreen();
        presenter.onQueryTextChange("");
        verify(mockView).showProgressBar(false);
    }

    @Test
    public void onQueryTextChangeShouldShowProgressBarWhenTextIsNotEmpty() {
        setupContactSelectionScreen();
        presenter.onQueryTextChange("1");
        verify(mockView).showProgressBar(true);
    }

    @Test
    public void onContactClickShouldDoNothingWhenShowingPreviewContactsScreen() {
        setupPreviewContactsScreen();
        presenter.onContactListUpdated(getMockContactList(false));

        presenter.onContactClick(mock(Contact.class));

        verify(mockView, never()).onContactSelected();
        verify(mockView, never()).showAddContestantButton(anyInt());
    }

    @Test
    public void onContactClickShouldNotifyAdapterDataSetChanged() {
        setupContactSelectionScreen();
        presenter.onContactListUpdated(getMockContactList(false));

        presenter.onContactClick(mock(Contact.class));

        verify(mockView).onContactSelected();
    }

    @Test
    public void onContactClickShouldShowAndIncreaseAddContestantButtonWhenUnselectedContactIsClicked() {
        setupContactSelectionScreen();
        List<Contact> list = getMockContactList(false);
        presenter.onContactListUpdated(list);

        presenter.onContactClick(list.get(0));

        verify(mockView).showAddContestantButton(1);
    }

    @Test
    public void onContactClickShouldShowAndDecreaseAddContestantButtonWhenOneOfMultipleSelectedContactsIsClicked() {
        setupContactSelectionScreen();
        List<Contact> list = getMockContactList(false);
        list.get(0).setSelected(true);
        list.get(1).setSelected(true);
        presenter.onContactListUpdated(list);

        presenter.onContactClick(list.get(0));

        verify(mockView).showAddContestantButton(1);
    }

    @Test
    public void onContactClickShouldHideAddContestantButtonWhenTheOnlySelectedContactIsClicked() {
        setupContactSelectionScreen();
        List<Contact> list = getMockContactList(false);
        list.get(0).setSelected(true);
        presenter.onContactListUpdated(list);

        presenter.onContactClick(list.get(0));

        verify(mockView).hideAddContestantButton();
    }

    @Test
    public void onAddParticipantsButtonClickedShouldShowSendInvitationsScreen() {
        setupContactSelectionScreen();
        presenter.onAddParticipantsButtonClicked();
        verify(mockView).showSendInvitationsScreen(any());
    }

    private List<Contact> getMockContactList(boolean selectOneValidContact) {
        final String EMPTY = "";
        final String TEST_PHONE = "555-555-5555";
        final String TEST_EMAIL = "email@test.com";
        List<Contact> contactList = new ArrayList<>();

        Contact onlyPhone = new Contact();
        onlyPhone.setId(1);
        onlyPhone.setPhone(TEST_PHONE);
        onlyPhone.setEmail(EMPTY);
        contactList.add(onlyPhone);
        if (selectOneValidContact) {
            onlyPhone.setSelected(true);
        }

        Contact onlyEmail = new Contact();
        onlyEmail.setId(2);
        onlyEmail.setPhone(EMPTY);
        onlyEmail.setEmail(TEST_EMAIL);
        contactList.add(onlyEmail);

        Contact phoneAndEmail = new Contact();
        phoneAndEmail.setId(3);
        phoneAndEmail.setPhone(TEST_PHONE);
        phoneAndEmail.setEmail(TEST_EMAIL);
        contactList.add(phoneAndEmail);

        Contact noPhoneOrEmail = new Contact();
        noPhoneOrEmail.setId(4);
        noPhoneOrEmail.setPhone(EMPTY);
        noPhoneOrEmail.setEmail(EMPTY);
        contactList.add(noPhoneOrEmail);

        return contactList;
    }
}

File: app/src/test/java/io/intrepid/contest/screens/sendinvitations/selectcontacts/SelectContactsPresenterTest.kt
package io.intrepid.contest.screens.sendinvitations.selectcontacts

import io.intrepid.contest.R
import io.intrepid.contest.models.Contact
import io.intrepid.contest.models.ParticipationType
import io.intrepid.contest.screens.sendinvitations.selectcontacts.SelectContactsContract.View
import io.intrepid.contest.testutils.BasePresenterTest
import junit.framework.Assert.assertTrue
import org.junit.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.ArgumentMatchers.anyList
import org.mockito.ArgumentMatchers.anyString
import org.mockito.ArgumentMatchers.argThat
import org.mockito.Mock
import org.mockito.Mockito.never
import org.mockito.Mockito.reset
import org.mockito.Mockito.verify
import java.util.*

class SelectContactsPresenterTest : BasePresenterTest<SelectContactsPresenter>() {
    @Mock
    private lateinit var mockView: View

    private fun setupContactSelectionScreen(participationType: ParticipationType = ParticipationType.CONTESTANT) {
        setup(participationType, true, ArrayList<Contact>())
    }

    private fun setupPreviewContactsScreen(participationType: ParticipationType) {
        setup(participationType, false, getMockContactList(true))
    }

    private fun setup(participationType: ParticipationType, contactSelectionEnabled: Boolean, contactList: List<Contact>) {
        presenter = SelectContactsPresenter(mockView,
                testConfiguration,
                participationType,
                contactSelectionEnabled,
                contactList)
    }

    @Test
    fun onViewCreatedShouldSetupAdapterWithContactSelectionWhenShowingContactSelectionScreen() {
        setupContactSelectionScreen()
        presenter.onViewCreated()
        verify<View>(mockView).setupAdapter(true)
    }

    @Test
    fun onViewCreatedShouldSetupAdapterWithoutContactSelectionWhenShowingPreviewContactsScreen() {
        setupPreviewContactsScreen(ParticipationType.CONTESTANT)
        presenter.onViewCreated()
        verify<View>(mockView).setupAdapter(false)
    }

    @Test
    fun onViewCreatedShouldDisplayPhoneContactListWhenShowingContactSelectionScreen() {
        setupContactSelectionScreen()
        presenter.onViewCreated()
        verify<View>(mockView).displayPhoneContactList()
    }

    @Test
    fun onViewCreatedShouldShowProgressBarWhenShowingContactSelectionScreen() {
        setupContactSelectionScreen()
        presenter.onViewCreated()
        verify<View>(mockView).showProgressBar(true)
    }

    @Test
    fun onViewCreatedShouldUpdateAdapterContactListWhenShowingPreviewContactsScreen() {
        setupPreviewContactsScreen(ParticipationType.CONTESTANT)
        presenter.onViewCreated()
        verify<View>(mockView).updateAdapterContactList(any<List<Contact>>())
    }

    @Test
    fun onViewCreatedShouldHideProgressBarWhenShowingPreviewContactsScreen() {
        setupPreviewContactsScreen(ParticipationType.CONTESTANT)
        presenter.onViewCreated()
        verify<View>(mockView).showProgressBar(false)
    }

    @Test
    fun onCreateOptionsMenuShouldCreateMenuSearchItemWhenShowingContactSelectionScreen() {
        setupContactSelectionScreen()
        presenter.onCreateOptionsMenu()
        verify<View>(mockView).createMenuSearchItem()
    }

    @Test
    fun onCreateOptionsMenuShouldNotCreateMenuSearchItemWhenShowingPreviewContactsScreen() {
        setupPreviewContactsScreen(ParticipationType.CONTESTANT)
        presenter.onCreateOptionsMenu()
        verify<View>(mockView, never()).createMenuSearchItem()
    }

    @Test
    fun onContactListUpdatedShouldHideProgressBar() {
        setupContactSelectionScreen()
        presenter.onContactListUpdated(getMockContactList(false))
        verify<View>(mockView).showProgressBar(false)
    }

    @Test
    fun onContactListUpdatedShouldUpdateContactListFilteringOutContactsWithoutPhoneOrEmail() {
        setupContactSelectionScreen()
        val initialContactList = getMockContactList(false)

        presenter.onContactListUpdated(initialContactList)

        verify<View>(mockView).updateAdapterContactList(
                argThat<List<Contact>> { argument -> argument.size == initialContactList.size - 1 })
    }

    @Test
    fun onContactListUpdatedShouldUpdateAdapterContactListKeepingDataFromContactWhenInExistedInPreviousList() {
        // Initial contact list
        setupContactSelectionScreen()
        val initialList = getMockContactList(false)
        initialList[0].isSelected = true
        presenter.onContactListUpdated(initialList)
        reset<View>(mockView)

        // Update list
        val listAfterSearch = getMockContactList(false).subList(0, 1)
        listAfterSearch[0].isSelected = false

        presenter.onContactListUpdated(listAfterSearch)

        verify<View>(mockView).updateAdapterContactList(
                argThat<List<Contact>> { adapterList -> adapterList[0] === initialList[0] })
    }

    @Test
    fun onQueryTextSubmitShouldAlwaysReturnTrue() {
        setupContactSelectionScreen()
        assertTrue(presenter.onQueryTextSubmit(""))
    }

    @Test
    fun onQueryTextChangeShouldUpdateContactSearchFilterWhenTextIsEmpty() {
        setupContactSelectionScreen()
        presenter.onQueryTextChange("")
        verify<View>(mockView).updateContactSearchFilter(anyString())
    }

    @Test
    fun onQueryTextChangeShouldNotUpdateContactSearchFilterWhenTextHasOneCharacter() {
        setupContactSelectionScreen()
        presenter.onQueryTextChange("1")
        verify<View>(mockView, never()).updateContactSearchFilter(anyString())
    }

    @Test
    fun onQueryTextChangeShouldUpdateContactSearchFilterWhenTextHasTwoCharactersOrMore() {
        setupContactSelectionScreen()
        presenter.onQueryTextChange("12")
        verify<View>(mockView).updateContactSearchFilter(anyString())
    }

    @Test
    fun onQueryTextChangeShouldHideProgressBarWhenTextIsEmpty() {
        setupContactSelectionScreen()
        presenter.onQueryTextChange("")
        verify<View>(mockView).showProgressBar(false)
    }

    @Test
    fun onQueryTextChangeShouldShowProgressBarWhenTextIsNotEmpty() {
        setupContactSelectionScreen()
        presenter.onQueryTextChange("1")
        verify<View>(mockView).showProgressBar(true)
    }

    @Test
    fun onContactClickShouldDoNothingWhenShowingPreviewContactsScreen() {
        setupPreviewContactsScreen(ParticipationType.CONTESTANT)
        presenter.onContactListUpdated(getMockContactList(false))

        presenter.onContactClick(Contact())

        verify<View>(mockView, never()).onContactSelected()
        verify<View>(mockView, never()).showAddContactsButton(anyInt(), anyInt())
    }

    @Test
    fun onContactClickShouldNotifyAdapterDataSetChanged() {
        setupContactSelectionScreen()
        val contactList = getMockContactList(false)
        presenter.onContactListUpdated(contactList)

        presenter.onContactClick(contactList[0])

        verify<View>(mockView).onContactSelected()
    }

    @Test
    fun onContactClickShouldShowAndIncreaseAddContestantsButtonWhenUnselectedContestantIsClicked() {
        setupContactSelectionScreen()
        val list = getMockContactList(false)
        presenter.onContactListUpdated(list)

        presenter.onContactClick(list[0])

        verify<View>(mockView).showAddContactsButton(1, R.plurals.numberOfContestants)
    }

    @Test
    fun onContactClickShouldShowAndIncreaseAddJudgesButtonWhenUnselectedJudgeIsClicked() {
        setupContactSelectionScreen(ParticipationType.JUDGE)
        val list = getMockContactList(false)
        presenter.onContactListUpdated(list)

        presenter.onContactClick(list[0])

        verify<View>(mockView).showAddContactsButton(1, R.plurals.numberOfJudges)
    }

    @Test
    fun onContactClickShouldShowAndDecreaseAddContestantsButtonWhenOneOfMultipleSelectedContestantsClicked() {
        setupContactSelectionScreen()
        val list = getMockContactList(false)
        list[0].isSelected = true
        list[1].isSelected = true
        presenter.onContactListUpdated(list)

        presenter.onContactClick(list[0])

        verify<View>(mockView).showAddContactsButton(1, R.plurals.numberOfContestants)
    }

    @Test
    fun onContactClickShouldShowAndDecreaseAddJudgesButtonWhenOneOfMultipleSelectedJudgesClicked() {
        setupContactSelectionScreen(ParticipationType.JUDGE)
        val list = getMockContactList(false)
        list[0].isSelected = true
        list[1].isSelected = true
        presenter.onContactListUpdated(list)

        presenter.onContactClick(list[0])

        verify<View>(mockView).showAddContactsButton(1, R.plurals.numberOfJudges)
    }

    @Test
    fun onContactClickShouldHideAddContactsButtonWhenTheOnlySelectedContactIsClicked() {
        setupContactSelectionScreen()
        val list = getMockContactList(false)
        list[0].isSelected = true
        presenter.onContactListUpdated(list)

        presenter.onContactClick(list[0])

        verify<View>(mockView).hideAddContactsButton()
    }

    @Test
    fun onAddParticipantsButtonClickedShouldShowSendInvitationsScreen() {
        setupContactSelectionScreen()
        presenter.onAddContactsButtonClicked()
        verify<View>(mockView).showSendInvitationsScreen(anyList<Contact>())
    }

    private fun getMockContactList(selectOneValidContact: Boolean): List<Contact> {
        val EMPTY = ""
        val TEST_PHONE = "555-555-5555"
        val TEST_EMAIL = "email@test.com"

        val onlyPhone = Contact().apply {
            id = 1
            phone = TEST_PHONE
            email = EMPTY
            isSelected = selectOneValidContact
        }

        val onlyEmail = Contact().apply {
            id = 2
            phone = EMPTY
            email = TEST_EMAIL
        }

        val phoneAndEmail = Contact().apply {
            id = 3
            phone = TEST_PHONE
            email = TEST_EMAIL
        }

        val noPhoneOrEmail = Contact().apply {
            id = 4
            phone = EMPTY
            email = EMPTY
        }

        return listOf(onlyPhone, onlyEmail, phoneAndEmail, noPhoneOrEmail)
    }
}
